compiler/rustc_typeck/src/check/generator_interior.rs

```diff
@@ -210,19 +226,49 @@ impl<'a, 'tcx> Visitor<'tcx> for InteriorVisitor<'a, 'tcx> {
         NestedVisitorMap::None
     }
 
+    fn visit_arm(&mut self, arm: &'tcx Arm<'tcx>) {
+        if arm.guard.is_some() {
+            self.nested_scope_of_guards.push(<_>::default());
+            self.arm_has_guard = true;
+        }
+        self.visit_pat(&arm.pat);
+        if let Some(ref g) = arm.guard {
+            match g {
+                Guard::If(ref e) => {
+                    self.visit_expr(e);
+                }
+            }
+            let mut scope_var_ids =
+                self.nested_scope_of_guards.pop().expect("should have pushed at least one earlier");
+            for var_id in scope_var_ids.drain(..) {
+                assert!(
+                    self.current_scope_of_guards.remove(&var_id),
+                    "variable should be placed in scope earlier"
+                );
+            }
+            self.arm_has_guard = false;
+        }
+        self.visit_expr(&arm.body);
+    }
+
     fn visit_pat(&mut self, pat: &'tcx Pat<'tcx>) {
         intravisit::walk_pat(self, pat);
 
         self.expr_count += 1;
 
-        if let PatKind::Binding(..) = pat.kind {
+        if let PatKind::Binding(_, id, ..) = pat.kind {
             let scope = self.region_scope_tree.var_scope(pat.hir_id.local_id);
             let ty = self.fcx.typeck_results.borrow().pat_ty(pat);
-            self.record(ty, Some(scope), None, pat.span);
+            self.record(ty, Some(scope), None, pat.span, false);
+            if self.arm_has_guard {
```

Comment:
We may want to check that this binding is _actually_ introduced by the if guard, instead of by some arbitrary expression inside it.

That said, it's okay if we overapproximate the set of types inside our generator. Since it's very rare to have other bindings introduced inside a match guard, we could leave this as is. If we do that we should leave a comment explaining that this is an overapproximation.

Reply:
I gave some thoughts about it. I think a better way is to use a separate and simple visitor that only walks the patterns in the arm, given that the arm has a guard. Most part of the original `visit_pat` will be kept the same.